File: Data Structures & Algorithms/burst-balloons/submission-2.cpp
class Solution {
private:
    int calc(vector<int>& nums, int li, int ri, int lb, int rb, unordered_map<string, int> &cache){
        if(li < 0 || ri >= nums.size() || li > ri){
            return 0;
        }

        if(li == ri){
            return nums[li] * lb * rb;
        }

        string key = to_string(li) + "-" + to_string(ri) + "-" + to_string(lb) + "-" + to_string(rb);

        if(cache.find(key) != cache.end()) return cache[key];
        int ans = 0;
        for(int j = li; j <= ri; j++){
            // calc left
            int left = calc(nums, li, j-1, lb, nums[j], cache);

            // calc right
            int right = calc(nums, j+1, ri, nums[j], rb, cache);

            ans = max(ans, nums[j]*lb*rb + left + right);
        }
        cache[key] = ans;
        return ans;
    }
public:
    int maxCoins(vector<int>& nums) {
        int n = nums.size();
        unordered_map<string, int> cache;
        return calc(nums, 0, n-1, 1, 1, cache);

    }
};

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] lp = new int[n];

        lp[0] = nums[0];
        for (int i = 1; i < n; i++) {
            lp[i] = lp[i - 1] * nums[i];
        }

        int rp = 1;
        for (int i = n - 1; i > 0; i--) {
            lp[i] = lp[i - 1] * rp;
            rp *= nums[i];
        }
        lp[0] = rp;
        return lp;
    }
}